apps/platform/src/journey/JourneyService.ts

```diff
@@ -1,93 +1,296 @@
     }
 }
 
-export const enterJourneyFromList = async (list: List, user: User, event?: UserEvent) => {
-    const steps = await JourneyStep.all(
-        qb => qb.leftJoin('journeys', 'journeys.id', 'journey_steps.journey_id')
-            .where('journeys.project_id', list.project_id)
-            .where('journeys.published', true)
-            .where('type', JourneyEntrance.type)
-            .whereJsonPath('data', '$.list_id', '=', list.id),
-    )
-    for (const step of steps) {
-        const service = new JourneyService(step.journey_id)
-        await service.run(user, event)
+export const enterAllUnstartedJourneysFromList = async (list: List) => {
+
+    // find all entrances that use this list
+    const entrances = await getActiveEntrancesForList(list)
+
+    if (!entrances.length) return
+
+    const ref = uuid()
+
+    const byJourneyId = groupBy(entrances, js => js.journey_id)
+
+    for (const [journey_id, [entrance]] of byJourneyId) {
+
+        // all users who have never started this journey
+        const query = UserList.query()
+            .select('user_lists.user_id as \'user_id\'')
+            .leftJoin('journey_user_step', 'user_lists.user_id', '=', 'journey_user_step.user_id')
+            .where('user_lists.list_id', list.id)
+            .where(qb => qb.whereNull('journey_user_step.user_id').orWhere('journey_user_step.journey_id', journey_id))
+            .groupBy('user_lists.user_id')
+            .having(UserList.raw('count(journey_user_step.id) = 0'))
+
+        await chunk<{ user_id: number }>(query, 100, async items => {
+            if (!items.length) return
+            await JourneyUserStep.insert(items.map(({ user_id }) => ({
+                journey_id,
+                user_id,
+                step_id: entrance.id,
+                type: 'completed',
+                ref,
+            })))
+        })
     }
+
+    return ref
 }
 
-export default class JourneyService {
+type JobOrJobFunc = Job | ((state: JourneyState) => Job)
+
+export class JourneyState {
+
+    /**
+     * resumes journey sequence/cycle processing for a given entrance (user can have multiple entrances, be in the journey multiple times)
+     * @param entrance entrance user step
+     * @param user target user to run journey for
+     * @returns promise that resolves when processing ends
+     */
+    public static async resume(entrance: number | JourneyUserStep, user?: User) {
+
+        // find entrance
+        if (typeof entrance === 'number') {
+            entrance = (await JourneyUserStep.find(entrance))!
+        }
+        if (!entrance) {
+            return
+        }
+        if (entrance.entrance_id) {
+            entrance = (await JourneyUserStep.find(entrance.entrance_id))!
+            if (!entrance || entrance.entrance_id) {
+                return
+            }
+        }
+
+        // entrance has already ended
+        if (entrance.ended_at) {
+            return
+        }
```

Comment:
This feels like it could be its own method, would make everything a bit more succinct as well